### PR TITLE
add DecAtt model

### PR DESCRIPTION
**Pull request by Victor0118**
Reference:
code: https://github.com/lanwuwei/SPM_toolkit/tree/master/DecAtt
paper: A Decomposable Attention Model for Natural Language Inference

@likicode @daemon Could you take a look at this PR when you are available?

### CODE REVIEW

**Comment by Victor0118** [comment]
==> decatt_sick_5e-4_0.001_0.5_0.1.log <==
INFO -          pearson_r       spearman_r      mse     KL-divergence loss
INFO - test     0.80094564      0.7184082390455326      0.3711671233177185      0.6171465432980202
DecAtt after tuning

File: decatt/model.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
+import sys
+import math
+import torch
+import numpy as np
+import torch.nn as nn
+import torch.nn.functional as F
```
Comment (daemon) [inline decatt/model.py]: PEP8 import order.

File: decatt/model.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+        :param num_classes: number of classes in the problem
+        :param embedding_size: size of each word embedding
+        :param use_intra_attention: whether to use intra-attention model
+        :param training: whether to create training tensors (optimizer)
+        :p/word_embeddingaram project_input: whether to project input embeddings to a
+            different dimensionality
+        :param distance_biases: number of different distances with biases used
+            in the intra-attention model
+        """
+        super(DecAtt, self).__init__()
```
Comment (daemon) [inline decatt/model.py]: `super().__init__()` is sufficient for Python 3.

File: decatt/model.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+        self.project_input = project_input
+        self.embedding_size=embedding_size
+        self.distance_biases=distance_biases
+        self.intra_attention=False
+        self.max_sentence_length=max_sentence_length
+        self.device = device
+
+        self.bias_embedding=nn.Embedding(max_sentence_length,1)
+        # self.word_embedding=nn.Embedding(vocab_size,embedding_size)
+        #self.linear_layer_project = Variable(torch.Tensor(embedding_size, projected_embedding_size))
```
Comment (daemon) [inline decatt/model.py]: Extraneous comments.

File: decatt/model.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+        self.linear_layer_compare[1].bias.data.fill_(0)
+        self.linear_layer_compare[4].weight.data.normal_(0, 0.01)
+        self.linear_layer_compare[4].bias.data.fill_(0)
+        self.linear_layer_aggregate[1].weight.data.normal_(0, 0.01)
+        self.linear_layer_aggregate[1].bias.data.fill_(0)
+        self.linear_layer_aggregate[4].weight.data.normal_(0, 0.01)
+        self.linear_layer_aggregate[4].bias.data.fill_(0)
+        #self.word_embedding.weight.data.copy_(torch.from_numpy(self.pretrained_emb))
+
+    def attention_softmax3d(self,raw_attentions):
```
Comment (daemon) [inline decatt/model.py]: Space between comma.

File: decatt/model.py [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+        reshaped_attentions = raw_attentions.view(-1, raw_attentions.size(2))
+        out=nn.functional.softmax(reshaped_attentions, dim=1)
+        return out.view(raw_attentions.size(0),raw_attentions.size(1),raw_attentions.size(2))
+
+    def _transformation_input(self,embed_sent):
+        #embed_sent=torch.matmul(embed_sent,self.linear_layer_project)
+        #embed_sent=self.word_embedding(embed_sent)
```
Comment (daemon) [inline decatt/model.py]: Extraneous comments.

File: decatt/model.py [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+        if self.intra_attention:
+            f_intra = self.linear_layer_intra(embed_sent)
+            f_intra_t = torch.transpose(f_intra, 1, 2)
+            raw_attentions = torch.matmul(f_intra, f_intra_t)
+            time_steps=embed_sent.size(1)
+            r = torch.arange(0, time_steps)
+            r_matrix=r.view(1,-1).expand(time_steps,time_steps)
+            raw_index=r_matrix-r.view(-1,1)
+            clipped_index=torch.clamp(raw_index,0,self.distance_biases-1)
+            clipped_index=Variable(clipped_index.long())
```
Comment (daemon) [inline decatt/model.py]: Space around the equal signs.

**Comment by daemon** [comment]
Logic seems fine, just nitpicks. Do we have results for *QA now?

**Comment by Victor0118** [comment]
I haven't run *QA. Might do that later.

**Comment by Victor0118** [comment]
@daemon All of your comments fixed.

With only one trial, the results of DecAtt on 
1. WikiQA is 4 points better than https://arxiv.org/pdf/1806.04330.pdf
2. TrecQA is 1 point lower than https://arxiv.org/pdf/1806.04330.pdf

**Comment by daemon** [comment]
Cool. There are still some extraneous comments, but I think it's in good enough shape.
